pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+			// This is not optimized yet - we should detect what changed based on event and only
+			// push resources that need to be pushed.
+
+			// TODO: possible race condition: if a config change happens while the envoy
+			// was getting the initial config, between LDS and RDS, the push will miss the
+			// monitored 'routes'. Same for CDS/EDS interval.
+			// It is very tricky to handle due to the protocol - but the periodic push recovers
+			// from it.
+
+			err := s.pushConnection(con, pushEv)
+			if err != nil {
+				return nil
+			}
+
+		}
+	}
+}
+
+/*func mergeSubscribeUnsubsribe(resource []string, subscribe []string, unsubscribe []string) []string {
+	for i := 0; i < len(resource); i++ {
+		for j := 0; j < len(subscribe); j++ {
+			for k := 0; k < len(unsubscribe); k++ {
+
+			}
+		}
+	}
+}*/
+
+// Send with timeout
+func (conn *XdsConnection) sendDelta(res *xdsapi.DeltaDiscoveryResponse) error {
+	done := make(chan error)
+	t := time.NewTimer(SendTimeout)
+	go func() {
+		err := conn.deltaStream.Send(res)
+		done <- err
+		conn.mu.Lock()
+		if res.Nonce != "" {
+			conn.RouteNonceSent = res.Nonce
+		}
+		conn.mu.Unlock()
+	}()
+	select {
+	case <-t.C:
+		// TODO: wait for ACK
+		adsLog.Infof("Timeout writing %s", conn.ConID)
+		xdsResponseWriteTimeouts.Add(1)
+		return errors.New("timeout sending")
+	case err := <-done:
+		t.Stop()
+		return err
+	}
+}
+
+func subscribeToVHost(sidecarScope *model.SidecarScope, pushContext *model.PushContext,
+routeConfigName uint32, host string, namespace string) (*model.SidecarScope, error) {
+	if sidecarScope == nil {
+		return nil, fmt.Errorf("nil sidecar scope")
+	}
+
+	var sidecarConfig *networking.Sidecar
+	var ok bool
+
+	if sidecarScope.Config != nil {
+		sidecarCfg := sidecarScope.Config
+		sidecarConfig, ok = sidecarCfg.Spec.(*networking.Sidecar)
+		if !ok {
+			return nil, fmt.Errorf("invalid sidecar config")
+		}
+	} else {
+		sidecarConfig = &networking.Sidecar{
+			Egress: []*networking.IstioEgressListener{
+				{
+					//Bavery_QUESTION: This is likely wrong, but how do we figure out the port info? Specifically the protocol
+					Port: &networking.Port{
+						Number:   routeConfigName,
+						Protocol: "HTTP",
+					},
+				},
+			},
+		}
+
+	}
+
+	found := false
+
+	for _, egressPb := range sidecarConfig.Egress {
+		if egressPb.Port.Number != routeConfigName {
+			continue
+		}
+		found = true
+		egressPb.Hosts = append(egressPb.Hosts, host)
+		break
+	}
+	if !found {
+		sidecarConfig.Egress = append(sidecarConfig.Egress, &networking.IstioEgressListener{
+			Port: &networking.Port{
+				Number:   routeConfigName,
+				Protocol: "HTTP",
+			},
+		})
+	}
+
+	sdCfg := &model.Config{
+		Spec: sidecarConfig,
+	}
+	updatedSidecarScope := model.ConvertToSidecarScope(pushContext, sdCfg, namespace)
+	return updatedSidecarScope, nil
+}
+
+//splitVHost returns the vhost and route config from a string
+func splitVHost(entry string) (uint32, string, error) {
+	vhostSeparator := "$"
```

Comment:
Where does this come from ? Protocol ? $ seems like a strange separator.

Reply:
$ is just a filler that I was using as a separator until we decided on one. Any suggestions for a better one?